### DISCUSSION TITLE
Problems with `.transform` and generics

### DISCUSSION THREAD

**Discussion by stuartkeith** [colinhacks/zod#1972] [Q&A] [answered]
Hi, I'm having trouble using `.transform` inside a function where Zod is using a generic passed into that function. For example (note this is a contrived example, not my actual use case):

```ts
function helper<Type extends string>(type: Type) {
    return z
        .object({
            type: z.literal(type),
            foo: z.number(),
        })
        .transform((obj) => {
            return {
                // does not work - obj.type does not exist on `obj`
                type: obj.type,
                // this is fine
                fooDoubled: obj.foo * 2,
            };
        });
}

const Hello = helper("hello");
const Goodbye = helper("goodbye");
```

The type of `obj` inside `transform` is:

```ts
obj: { [k_1 in keyof z.objectUtil.addQuestionMarks<{
    type: Type;
    foo: number;
}>]: z.objectUtil.addQuestionMarks<{
    type: Type;
    foo: number;
}>[k_1]; }
```

`obj.type` is not accessible on the object at all. I can see the expected `type: Type` generic is there but it appears that the `addQuestionMarks` type is losing the generic.

This also happens if I use `Type extends z.ZodLiteral<string>` and pass that in directly.

I'm not sure if this is a known limitation of zod or a bug/missing feature. Is there any way of getting around this? Thanks.

**Comment by JacobWeisenburger** [comment, marked as answer]
moved to issue: https://github.com/colinhacks/zod/issues/1977

**Comment by stuartkeith** [comment]
Seems to be fixed in zod 4, thanks!
